tests/test-selection-transfer.lua

```diff
@@ -0,0 +1,235 @@
+        return true
+    end,
+
+    function()
+        -- Wait for the test to succeed
+        if not continue then
+            return
+        end
+        continue = false
+
+        -- Now test piece-wise selection transfers
+        selection = assert(selection_acquire{ selection = "CLIPBOARD" },
+            "Failed to acquire the clipboard selection")
+        selection:connect_signal("request", function(_, target, transfer)
+            if target == "TARGETS" then
+                transfer:send{
+                    format = "atom",
+                    data = { "TARGETS", "UTF8_STRING" },
+                }
+            elseif target == "UTF8_STRING" then
+                local offset = 1
+                local data = "Hello World!"
+                local function send_piece()
+                    local piece = data:sub(offset, offset)
+                    transfer:send{
+                        data = piece,
+                        continue = piece ~= ""
+                    }
+                    offset = offset + 1
+                end
+                transfer:connect_signal("continue", send_piece)
+                send_piece()
+            end
+        end)
+        awesome.sync()
+        spawn.with_line_callback({ "lua", "-e", check_targets_and_text },
+            { stdout = function(line)
+                assert(line == "done", "Unexpected line: " .. line)
+                continue = true
+            end })
+        return true
+    end,
+
+    function()
+        -- Wait for the test to succeed
+        if not continue then
+            return
+        end
+        continue = false
+
+        -- Now test a huge transfer
+        selection = assert(selection_acquire{ selection = "CLIPBOARD" },
+            "Failed to acquire the clipboard selection")
+        selection:connect_signal("request", function(_, target, transfer)
+            if target == "TARGETS" then
+                transfer:send{
+                    format = "atom",
+                    data = { "TARGETS", "UTF8_STRING" },
+                }
+            elseif target == "UTF8_STRING" then
+                local count = 0
+                local function send_piece()
+                    count = count + 1
+                    local done = count == large_transfer_piece_count
+                    transfer:send{
+                        data = large_transfer_piece,
+                        continue = not done,
+                    }
+                end
+                transfer:connect_signal("continue", send_piece)
+                send_piece()
+            end
+        end)
+        awesome.sync()
+        spawn.with_line_callback({ "lua", "-e", check_large_transfer },
+            { stdout = function(line)
+                assert(line == "done", "Unexpected line: " .. line)
+                continue = true
+            end })
+        return true
+    end,
+
+    wait_a_bit,
+    wait_a_bit,
+    wait_a_bit,
+    wait_a_bit,
+    wait_a_bit,
```

Comment:
Why are those `wait_a_bit` calls necessary?  Could have a comment maybe, if necessary.

Reply:
Done in a new commit.

Because transferring 100 MiB of data (inefficiently) is slow. I had a timeout without `wait_a_bit` (which means that it took longer than 0.5 seconds, I think). I just added a couple of these calls because others might have slower computers than I do.

The previous step tests large data transfer. The goal here is to have a single piece of data that is so large that the C code has to split it up into pieces, too. Thus, a single `:send` turns into multiple `INCR` steps for the C code. This also tests continuation, i.e. Lua adding more data later. If my math is correct, this almost transfers 100 MiB of data through the X11 server.

The maximum request size of my current X11 server is a bit below 16 MiB. Thus, each of the 32 MiB pieces that Lua gives to C has to split up into three increments, two almost of full size and one with just 8 bytes of data. :-)